### PR TITLE
grpclb: Don't go into fallback if we haven't started picking

### PR DESCRIPTION
**Pull request by dgquintas**
Also includes improvements to logging that help while debugging (always print the address of the grpclb policy, and in a consistent manner).

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                              FILE SIZE
 ++++++++++++++ GROWING                                                                ++++++++++++++
  +0.0%    +160 [None]                                                                 +3.44Ki  +0.1%
  +0.4%     +64 src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc             +64  +0.4%
      +8.8%     +32 [Unmapped]                                                                 +32  +8.8%
      +1.7%     +27 rr_handover_locked(grpc_exec_ctx*, glb_lb_policy*) [clone .part.5]         +27  +1.7%
      +2.0%     +27 lb_on_response_received_locked                                             +27  +2.0%
      +2.9%     +19 pick_from_internal_rr_locked                                               +19  +2.9%
      +5.3%     +16 update_lb_connectivity_status_locked                                       +16  +5.3%
      +0.3%      +3 glb_create                                                                  +3  +0.3%
      +0.5%      +3 send_client_load_report_locked                                              +3  +0.5%
      +0.7%      +3 lb_on_server_status_received_locked                                         +3  +0.7%

  +0.0%    +224 TOTAL                                                                  +3.50Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

File: src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc [hunk 1]
```diff
@@ -619,8 +623,10 @@ static void update_lb_connectivity_status_locked(
 
   if (GRPC_TRACER_ON(grpc_lb_glb_trace)) {
     gpr_log(
-        GPR_INFO, "Setting grpclb's state to %s from new RR policy %p state.",
-        grpc_connectivity_state_name(rr_state), (void*)glb_policy->rr_policy);
+        GPR_INFO,
+        "[GLB %p] Setting grpclb's state to %s from new RR policy %p state.",
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc]: Instead of "GLB", let's use "grpclb" in the log messages.  I think it will be easier to read for the person reading the logs.

**Comment by markdroth** [comment]
Also, is there a reasonable way to add a test for this?

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                              FILE SIZE
 ++++++++++++++ GROWING                                                                ++++++++++++++
  +0.0%    +160 [None]                                                                 +3.44Ki  +0.1%
  +0.4%     +64 src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc             +64  +0.4%
      +8.8%     +32 [Unmapped]                                                                 +32  +8.8%
      +1.7%     +27 rr_handover_locked(grpc_exec_ctx*, glb_lb_policy*) [clone .part.5]         +27  +1.7%
      +2.0%     +27 lb_on_response_received_locked                                             +27  +2.0%
      +2.9%     +19 pick_from_internal_rr_locked                                               +19  +2.9%
      +5.3%     +16 update_lb_connectivity_status_locked                                       +16  +5.3%
      +0.3%      +3 glb_create                                                                  +3  +0.3%
      +0.5%      +3 send_client_load_report_locked                                              +3  +0.5%
      +0.7%      +3 lb_on_server_status_received_locked                                         +3  +0.7%

  +0.0%    +224 TOTAL                                                                  +3.50Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by dgquintas** [comment]
I've spent a couple hours trying to come up with a simple way to test this, but failed. Basically what needs to happen is:
1. grpclb policy is created
2. fallback timer fires, creates a RR over the backends because `serverlist == NULL`
3. a `pick()` happens. Previously it'd assert on `client_stats != NULL`, but `client_stats` is set in `lb_call_init()`, but that hasn't happened because `pick()` only calls it when RR is NULL, which is isn't on account of point 2.  

Replicating this in a high level test would require the ability to separate the LB creation from the `pick()`, something we don't have from the testing/application layer.

**Comment by markdroth** [comment]
Hmmm... Then how is this crash actually happening in production?  Looking at the code, I don't actually see how it's possible.  We trigger both starting the fallback timer and setting `glb_policy->client_stats` from the same place, `start_picking_locked()`.  So I don't see how `client_stats` is not set when the pick comes in.

**Comment by dgquintas** [comment]
Let me try again, because the process doesn't involve the fallback timer firing:

1. grpclb policy is created
2. *a resolver update comes*: `glb_update_locked()` is invoked. Because `glb_policy->serverlist == NULL`, we call `fallback_update_locked()`, which will call `rr_handover_locked()` because the fallback timer is active (NOT fired yet, as I previously stated incorrectly). This will create a fallback-RR instance.
3. a `pick()` happens. Previously it'd assert on `client_stats != NULL`, but `client_stats` is set in `lb_call_init()`, but that hasn't happened because `pick()` only calls it when RR is NULL, which is isn't on account of point 2.  

With this in mind, I may be able to simulate this in a test by having a separate thread sending a deluge of naming updates very frequently. This, to be honest, seems like a good thing to do for stress testing in general. I'll try this approach tomorrow.

**Comment by AspirinSJL** [comment]
Regarding to point 2: `rr_handover_locked()` will only be called when `!glb_policy->fallback_timer_active`, which means the timer has already been fired.

Regarding to point 3: `lb_call_init_locked()` should have been invoked before the fallback timer fires, because the fallback timer is started only in `start_picking_locked()` and `start_picking_locked()` calls `query_for_backends_locked()` which calls `lb_call_init_locked()`.

**Comment by dgquintas** [comment]
I've edited point 2: the timer is not active, but not because the timer has already fired, but because it hasn't started yet.

**Comment by dgquintas** [comment]
PTAL. I've re-done the commits, added yet another fix, and changed the approach for `client_stats`, as suggested.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                     FILE SIZE
 ++++++++++++++ GROWING                                                                       ++++++++++++++
  +1.3%    +192 src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc                   +192  +1.3%
      +9.8%     +63 pick_from_internal_rr_locked                                                      +63  +9.8%
      +3.0%     +39 query_for_backends_locked(grpc_exec_ctx*, glb_lb_policy*) [clone .part.7]         +39  +3.0%
      +1.7%     +27 rr_handover_locked(grpc_exec_ctx*, glb_lb_policy*) [clone .part.5]                +27  +1.7%
      +2.0%     +27 lb_on_response_received_locked                                                    +27  +2.0%
      +5.3%     +16 update_lb_connectivity_status_locked                                              +16  +5.3%
      +3.8%     +14 [Unmapped]                                                                        +14  +3.8%
      +0.3%      +3 glb_create                                                                         +3  +0.3%
      +0.5%      +3 send_client_load_report_locked                                                     +3  +0.5%
      +0.7%      +3 lb_on_server_status_received_locked                                                +3  +0.7%

 -+-+-+-+-+-+-+ MIXED                                                                         +-+-+-+-+-+-+-
  +0.1%    +288 [None]                                                                           -552  -0.0%

  +0.0%    +480 TOTAL                                                                            -360  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

File: src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc [hunk 2]
```diff
@@ -202,7 +202,7 @@ static void wrapped_rr_closure(grpc_exec_ctx* exec_ctx, void* arg,
         gpr_log(GPR_ERROR,
                 "No LB token for connected subchannel pick %p (from RR "
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc]: Should we include the address of the grpclb policy in this log message?

File: src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc [hunk 3]
```diff
@@ -213,7 +213,7 @@ static void wrapped_rr_closure(grpc_exec_ctx* exec_ctx, void* arg,
       grpc_grpclb_client_stats_unref(wc_arg->client_stats);
     }
     if (GRPC_TRACER_ON(grpc_lb_glb_trace)) {
-      gpr_log(GPR_INFO, "Unreffing RR %p", (void*)wc_arg->rr_policy);
+      gpr_log(GPR_INFO, "Unreffing RR %p", wc_arg->rr_policy);
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc]: Should we include the address of the grpclb policy in this log message?

File: src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc [hunk 4]
```diff
@@ -647,15 +649,16 @@ static bool pick_from_internal_rr_locked(
     if (server->drop) {
       // Not using the RR policy, so unref it.
       if (GRPC_TRACER_ON(grpc_lb_glb_trace)) {
-        gpr_log(GPR_INFO, "Unreffing RR for drop (0x%" PRIxPTR ")",
-                (intptr_t)wc_arg->rr_policy);
+        gpr_log(GPR_INFO, "[grpclb %p] Unreffing RR for drop (0x%" PRIxPTR ")",
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc]: Nit: Can use `%p` for the RR policy address.

File: src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc [hunk 5]
```diff
@@ -676,7 +679,7 @@ static bool pick_from_internal_rr_locked(
   if (pick_done) {
     /* synchronous grpc_lb_policy_pick call. Unref the RR policy. */
     if (GRPC_TRACER_ON(grpc_lb_glb_trace)) {
-      gpr_log(GPR_INFO, "Unreffing RR (0x%" PRIxPTR ")",
+      gpr_log(GPR_INFO, "[grpclb %p] Unreffing RR (0x%" PRIxPTR ")", glb_policy,
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc]: Nit: Can use `%p` for the RR policy address.

File: src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc [hunk 6]
```diff
@@ -1752,18 +1768,18 @@ static void glb_update_locked(grpc_exec_ctx* exec_ctx, grpc_lb_policy* policy,
...
+          glb_policy);
     }
     return;
   }
   const grpc_lb_addresses* addresses =
       (const grpc_lb_addresses*)arg->value.pointer.p;
   // If a non-empty serverlist hasn't been received from the balancer,
   // propagate the update to fallback_backend_addresses.
-  if (glb_policy->serverlist == NULL) {
+  if (glb_policy->started_picking && glb_policy->serverlist == NULL) {
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc]: I don't think this is correct.

Two things happen inside of `fallback_update_locked()`:
1. We update `glb_policy->fallback_backend_addresses` with the new list of backends provided by the resolver.
2. If we are already in fallback mode, we pass that updated list of addresses to the RR policy by calling `rr_handover_locked()`.

Note that even if we have not yet started picking, we still need to do (1), so that when we do start picking later, we will use the right set of backend addresses as the fallback.  However, we do not want (2) to happen in this case.  So I think the bug here is that `fallback_update_locked()` is using the wrong condition to decide when to call `rr_handover_locked()`.  This fix should be made there instead of here.

Note that in #12587, I suggested that the condition in `fallback_update_locked()` should actually be `glb_policy->rr_policy != NULL`.  However, @AspirinSJL responded correctly that that causes problems when the RR policy shuts down due to all of the backends going away.  That problem will go away with #12829, so if that PR gets merged before this one, we can just check that.

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

File: src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc [hunk 7]
```diff
@@ -1482,13 +1486,15 @@ static void query_for_backends_locked(grpc_exec_ctx* exec_ctx,
   GPR_ASSERT(glb_policy->lb_channel != NULL);
   if (glb_policy->shutting_down) return;
 
+  if (glb_policy->retry_timer_active) {
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc]: What problem does this fix?  In what case can we get here when the retry timer is active?

If this fix is unrelated to the `client_stats` fix, it might be better to move it to a separate PR.

Comment (dgquintas) [reply src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc]: This:

```
I1109 09:40:39.169257  994714 grpclb.cc:1495] [grpclb 0x7f1dbb048300] Query for backends (lb_channel: 0x7f1dba165000, lb_call: 0x7f1db5480020)
...
I1109 09:40:39.171471  994714 grpclb.cc:1267] [grpclb 0x7f1dbb048300] Restaring call to LB server
E1109 09:40:39.172154  994714 grpclb.cc:1269] assertion failed: glb_policy->lb_call == NULL

```
That is, the timer, having a life of its own, may fire after the LB call has already been restarted.

ok, I'll move it to another PR.

File: src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc (hunk 2 above)
Comment (dgquintas) [reply src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc]: `glb_policy` isn't available here. We could add it to `wc_args`...

File: src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc (hunk 3 above)
Comment (dgquintas) [reply src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc]: Same as in the other comment, `glb_policy` not available. Possible to make it, but it'd only be used for logging. Still, it may be worthwhile, given how much easier it is to debug issues with good logging.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                              FILE SIZE
 ++++++++++++++ GROWING                                                                ++++++++++++++
  +1.1%    +160 src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc            +160  +1.1%
      +9.8%     +63 pick_from_internal_rr_locked                                               +63  +9.8%
      +1.7%     +27 rr_handover_locked(grpc_exec_ctx*, glb_lb_policy*) [clone .part.5]         +27  +1.7%
      +2.0%     +27 lb_on_response_received_locked                                             +27  +2.0%
      +5.8%     +21 [Unmapped]                                                                 +21  +5.8%
      +5.3%     +16 update_lb_connectivity_status_locked                                       +16  +5.3%
      +0.3%      +3 glb_create                                                                  +3  +0.3%
      +0.5%      +3 send_client_load_report_locked                                              +3  +0.5%
      +0.7%      +3 lb_on_server_status_received_locked                                         +3  +0.7%

 -+-+-+-+-+-+-+ MIXED                                                                  +-+-+-+-+-+-+-
  +0.1%    +288 [None]                                                                    -232  -0.0%

  +0.0%    +448 TOTAL                                                                      -72  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

File: src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc (hunk 6 above)
Comment (dgquintas) [reply src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc]: I see. I'm dropping this PR. I've separated the other two concerns in https://github.com/grpc/grpc/pull/13332 (logging) and https://github.com/grpc/grpc/pull/13331 (retry timer).

Comment (dgquintas) [reply src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc]: I've filed https://github.com/grpc/grpc/issues/13333 to track this.

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc (hunk 6 above)
Comment (markdroth) [reply src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc]: Perhaps I was unclear.  I don't think we need a separate issue for this, nor do I think we need to wait for #12829 to go in first.  I think we can just change `fallback_update_locked()` from this:

```
   if (glb_policy->lb_fallback_timeout_ms > 0 &&
       !glb_policy->fallback_timer_active) {
```

to this:

```
   if (glb_policy->started_picking &&
       glb_policy->lb_fallback_timeout_ms > 0 &&
       !glb_policy->fallback_timer_active) {
```

I think we can go ahead and do that immediately, without waiting for #12829.

My comment was simply that once #12829 goes in, the condition in `fallback_update_locked()` can change from what's shown above to simply `glb_policy->rr_policy != NULL`.

Comment (AspirinSJL) [reply src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc]: I have appended a commit in #12829 to change the condition to "glb_policy->rr_policy != NULL".
